File: sort-an-array/Solution912.java
// https://leetcode.com/problems/sort-an-array/

import java.util.Arrays;

class Solution912 {

    // Merge sort
    // T: O(n log n), S: O(n)
    public int[] sortArray(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return nums;
        }
        var len = nums.length;
        var mid = len / 2;
        var nums1 = new int[mid];
        var nums2 = new int[len - mid];
        System.arraycopy(nums, 0, nums1, 0, mid);
        System.arraycopy(nums, mid, nums2, 0, len - mid);
        sortArray(nums1);
        sortArray(nums2);
        merge(nums, nums1, nums2);

        return nums;
    }

    private void merge(int[] nums, int[] nums1, int[] nums2) {
        int len1 = nums1.length, len2 = nums2.length;
        int idx = 0, idx1 = 0, idx2 = 0;
        while (idx1 < len1 && idx2 < len2) {
            nums[idx++] = nums1[idx1] <= nums2[idx2] ? nums1[idx1++] : nums2[idx2++];
        }
        while (idx1 < len1) {
            nums[idx++] = nums1[idx1++];
        }
        while (idx2 < len2) {
            nums[idx++] = nums2[idx2++];
        }
    }

    public static void main(String[] args) {
        var sol = new Solution912();
        var nums1 = new int[] { 5, 2, 3, 1 };
        sol.sortArray(nums1);
        System.out.println(Arrays.toString(nums1));
        var nums2 = new int[] { 5, 1, 1, 2, 0, 0 };
        sol.sortArray(nums2);
        System.out.println(Arrays.toString(nums2));
    }

}

File: sort-an-array/solution912.py
# https://leetcode.com/problems/sort-an-array/

from typing import List


class Solution912:

    # Merge sort
    # T: O(n log n), S: O(n)
    def sortArray(self, nums: List[int]) -> List[int]:
        if not nums or len(nums) <= 1:
            return nums

        mid = len(nums) // 2
        nums1 = nums[:mid]
        nums2 = nums[mid:]

        self.sortArray(nums1)
        self.sortArray(nums2)
        self.merge(nums, nums1, nums2)

        return nums

    def merge(self, nums: List[int], nums1: List[int], nums2: List[int]) -> None:
        idx = idx1 = idx2 = 0
        len1, len2 = len(nums1), len(nums2)

        while idx1 < len1 and idx2 < len2:
            if nums1[idx1] <= nums2[idx2]:
                nums[idx] = nums1[idx1]
                idx1 += 1
            else:
                nums[idx] = nums2[idx2]
                idx2 += 1
            idx += 1

        while idx1 < len1:
            nums[idx] = nums1[idx1]
            idx1 += 1
            idx += 1

        while idx2 < len2:
            nums[idx] = nums2[idx2]
            idx2 += 1
            idx += 1


def main() -> None:
    sol = Solution912()
    nums1 = [5, 2, 3, 1]
    print(sol.sortArray(nums1))
    nums2 = [5, 1, 1, 2, 0, 0]
    print(sol.sortArray(nums2))


if __name__ == "__main__":
    main()
